### DISCUSSION TITLE
Best Practices for Writing Effective GitHub Issues

### DISCUSSION THREAD

**Discussion by zufii5030** [community/community#147722] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hello everyone! 👋  

I've been working on my first collaborative project, and I've realized how important clear communication is when reporting bugs or suggesting features through GitHub Issues. However, I sometimes struggle to write issues that are concise yet detailed enough for my collaborators to understand and address efficiently.  

I wanted to ask:  
1. **What are the key elements of an effective GitHub Issue?**  
2. **Do you use specific templates or formats to ensure consistency? If so, could you share examples?**  
3. **How do you prioritize issues in your projects?**  
4. **Are there any tools or GitHub integrations you recommend to improve issue tracking and management?**  

I'd love to hear your experiences, tips, and any resources that have helped you!  

Thank you in advance for sharing your insights. 😊

**Comment by sulhanfuadi** [comment, marked as answer]
Hi there! 👋  

Great question! Writing effective GitHub Issues is a skill that can significantly improve collaboration and project efficiency. Here’s a detailed guide based on my experience:  

---  

### **1. Key Elements of an Effective GitHub Issue**  
To write a clear and actionable GitHub Issue, include:  
- **A Descriptive Title:** Summarize the issue concisely, e.g., "Bug: App crashes when clicking on the login button."  
- **Detailed Description:** Provide context, including what you expected to happen versus what actually occurred.  
- **Steps to Reproduce:** Include step-by-step instructions for replicating the bug.  
  Example:  
  1. Open the app.  
  2. Navigate to the login screen.  
  3. Click on the login button without entering credentials.  
- **Screenshots or Logs:** Visuals or logs can help others understand the problem quickly.  
- **Environment Details:** Mention the OS, browser, or software version in use, if relevant.  
- **Labels:** Use tags like "bug," "enhancement," or "question" to categorize the issue properly.  

---  

### **2. Templates for Consistency**  
Using GitHub Issue templates is a game-changer. Here’s a basic example:  

**Bug Report Template:**  
```  
### Description  
(Provide a clear and concise description of the problem.)  

### Steps to Reproduce  
1. [Step 1]  
2. [Step 2]  

### Expected Behavior  
(Explain what you expected to happen.)  

### Actual Behavior  
(Explain what actually happened.)  

### Environment  
- OS:  
- Browser/Version:  

### Additional Information  
(Add screenshots, logs, or other helpful details.)  
```  

GitHub allows you to configure these templates in the `.github/ISSUE_TEMPLATE` folder of your repository.  

---  

### **3. Prioritizing Issues**  
When managing issues, I use these techniques:  
- **Prioritize by Impact and Effort:** Use a simple matrix to classify issues into "Quick Wins," "Big Bets," etc.  
- **Label by Priority:** Create labels like `priority:high`, `priority:medium`, and `priority:low`.  
- **Milestones:** Assign issues to specific milestones to organize your release goals.  
- **Assignees:** Clearly designate who’s responsible for resolving each issue.  

---  

### **4. Recommended Tools and Integrations**  
- **ZenHub or GitHub Projects:** For Kanban-style issue tracking.  
- **Slack or Discord Integrations:** Automatically notify your team of new issues.  
- **Linear or Clubhouse:** Integrate with GitHub for enhanced project management.  
- **Dependabot:** Automatically manage dependency updates.  

---  

### **Final Tip:**  
Encourage collaborators to ask questions if something isn’t clear in an issue. Open communication fosters better teamwork and ensures issues are resolved efficiently.  

Hope this helps! Happy coding and collaborating. 🚀

**Comment by jayDevCodes** [comment]
Title: be concise and specific (what failed + where). Example: “API v2 — 500 on /login when payload missing field”.

Reproduce steps: list exact steps, minimal repro, expected vs actual behavior, and relevant environment (OS, browser, versions).

Attach logs, screenshots, and a small repro repo or Gist — maintainers appreciate reproducible examples.

Use labels and templates (if provided) and avoid broad “works on my machine” claims; say what you tried.

Close with a concise ask: “Expected: X. Actual: Y. Can you suggest a fix or cause?”
